### PR TITLE
Create table or proc should throw error for declaring non-permissible datalength in variable length datatypes; Fix tests related to Numeric datatype in REL3_X_DEV branch

### PR DESCRIPTION
**Pull request by bsatarupa**
### Description
PG15 has introduced a community change for numeric datatype as below:
Formerly, when specifying NUMERIC(precision, scale), the scale had to be in the range [0, precision], which was per SQL spec. This commit extends the range of allowed scales to [-1000, 1000], independent of
the precision (whose valid range remains [1, 1000]). A negative scale implies rounding before the decimal point. For
example, a column might be declared with a scale of -3 to round values to the nearest thousand. Note that the display scale remains non-negative, so in this case the display scale will be zero, and all digits before the decimal point will be displayed.
A scale greater than the precision supports fractional values with zeros immediately after the decimal point.

From SQL spec aspect which means checking explicitly whether 0 <= scale <= precision and 1 <= precision <= 38, which was earlier done by underlying PG code for any numeric/decimal input data.

Along with the above change, it is found that all variable length datatypes like numeric, decimal, time, datetime2, datetimeoffset need explicit checking on their permissible datalength at the time of table or stored procedure creation and generate error if required.
Before fix, Babelfish was not complaining for the above behaviour.

Task: BABEL-3720, BABEL-3793
Signed-off-by: Satarupa Biswas (satarupb@amazon.com)

### Test Scenarios Covered ###
* **Use case based -**  Addressed basic testcase failure that are part of PG15 related numeric changes

* **Boundary conditions -** Added testcases with scale and precision values inside and outside permissible range for variable length datatypes

* **Arbitrary inputs -** Added

* **Negative test cases -** Added testcases with negative scale and precision values

* **Minor version upgrade tests -** Added

* **Major version upgrade tests -** Added

* **Performance tests -** N/A

* **Tooling impact -** N/A

* **Client tests -** N/A

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 1]
```diff
@@ -275,6 +278,9 @@ InstallExtendedHooks(void)
 	planner_hook = pltsql_planner_hook;
 	prev_transform_check_constraint_expr_hook = transform_check_constraint_expr_hook;
 	transform_check_constraint_expr_hook = transform_like_in_add_constraint;
+
+	prev_verify_valid_scale_in_variable_length_datatypes_hook = verify_valid_scale_in_variable_length_datatypes_hook;
+	verify_valid_scale_in_variable_length_datatypes_hook = pltsql_verify_valid_scale_in_variable_length_datatypes;
```
Comment (rishabhtanwar29) [inline contrib/babelfishpg_tsql/src/hooks.c]: Does `validate_var_datatype_scale_hook` makes sense? (Trying to keep the name short)

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 2]
```diff
@@ -2987,3 +2994,50 @@ transform_like_in_add_constraint (Node* node)
 	
 	return pltsql_predicate_transformer(node);
 }
+
+/*
+ * pltsql_verify_valid_scale_in_variable_length_datatypes()
+ * - Checks whether variable length datatypes like numeric, decimal, time, datetime2, datetimeoffset 
+ * are declared with permissible datalength at the time of table or stored procedure creation
+ */
+void pltsql_verify_valid_scale_in_variable_length_datatypes(char *dataTypeName, int scale, int precision)
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/hooks.c]: is this the only solution which will work? Parsing data type name for decision making is not very robust. Can't we have some other data structure which can provide actual types (ENUM or equivalent ) for decision making

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 1 above)
Comment (bsatarupa) [reply contrib/babelfishpg_tsql/src/hooks.c]: done

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 2 above)
Comment (bsatarupa) [reply contrib/babelfishpg_tsql/src/hooks.c]: Engine side change: https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/pull/69

Inside parser function: static int32 typenameTypeMod(ParseState *pstate, const TypeName *typeName, Type typ); typeName and its scale is used to decide on valid input scale value, if any.

File: test/JDBC/upgrade/14_3/schedule [hunk 3]
```diff
@@ -290,3 +290,4 @@ BABEL-CHECK-CONSTRAINT-before-14_6
 BABEL-3640
 sys_server_principals_dep
 sys_database_principals_dep
+TestVariableDataLength
```
Comment (rishabhtanwar29) [inline test/JDBC/upgrade/14_3/schedule]: Let's not add this test here since the error messages are changing in 15 compared to 14.

File: test/JDBC/upgrade/13_5/schedule [hunk 4]
```diff
@@ -242,3 +242,4 @@ BABEL-3655
 sys-table_types_internal
 BABEL-CHECK-CONSTRAINT-before-14_6
 sys_server_principals_dep_for_13_x
+TestVariableDataLength
```
Comment (rishabhtanwar29) [inline test/JDBC/upgrade/13_5/schedule]: Let's not add this test here since the error messages are changing in 15 compared to 13.

File: test/JDBC/upgrade/14_3/schedule (hunk 3 above)
Comment (bsatarupa) [reply test/JDBC/upgrade/14_3/schedule]: Alright

File: test/JDBC/upgrade/13_5/schedule (hunk 4 above)
Comment (bsatarupa) [reply test/JDBC/upgrade/13_5/schedule]: Alright

File: test/JDBC/upgrade/latest/schedule [hunk 5]
```diff
@@ -49,9 +49,8 @@ BABEL-3221
 BABEL-3204
 BABEL-3234
 BABEL-3402
-# Disabled until BABEL-3720 gets fixed
```
Comment (rishabhtanwar29) [inline test/JDBC/upgrade/latest/schedule:52]: Make sure to remove these tests from 14_6/schedule as well since 14_6/ folder was created from latest/

Comment (bsatarupa) [reply test/JDBC/upgrade/latest/schedule:52]: not found in 14_6/schedule

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 6]
```diff
@@ -2987,3 +2994,75 @@ transform_like_in_add_constraint (Node* node)
...
+ */
+void pltsql_validate_var_datatype_scale(const TypeName *typeName, Type typ)
+{
+	Oid datatype_oid = InvalidOid;
+	int count = 0;
+	ListCell   *l;
+	int scale[2] = {-1, -1};
+
+	char *dataTypeName = TypeNameToString(typeName);
+	if (strstr(dataTypeName, "pg_catalog."))
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/hooks.c]: I think there is PG function which returns both schema and type name for given typeName. Please use that

File: test/JDBC/expected/BABEL-3486-vu-prepare.out [hunk 7]
```diff
@@ -1057,6 +1057,10 @@ GO
 
 CREATE VIEW BABEL_3486_vu_prepare_v102_2 as (SELECT TRY_CONVERT(datetime2(10), '2017-08-25 13:01:59.1234567'));
 GO
+~~ERROR (Code: 33557097)~~
+
+~~ERROR (Message: Specified scale 10 is invalid. 'datetime2' datatype must have scale between 0 and 7)~~
```
Comment (surendravishnoi) [inline test/JDBC/expected/BABEL-3486-vu-prepare.out:1062]: is this expected? what happens in case of SQL server? if it is not a valid SQL server use case, remove the test

File: test/JDBC/expected/BABEL-3614-vu-prepare.out [hunk 8]
```diff
@@ -44,6 +44,10 @@ GO
 
 CREATE VIEW BABEL_3614_vu_prepare_v5 as (SELECT TRY_CONVERT(DATETIME2(80), '2017-08-25 13:01:10.1234567'));
 GO
+~~ERROR (Code: 33557097)~~
+
+~~ERROR (Message: Specified scale 80 is invalid. 'datetime2' datatype must have scale between 0 and 7)~~
```
Comment (surendravishnoi) [inline test/JDBC/expected/BABEL-3614-vu-prepare.out:49]: is this expected? what happens in case of SQL server?

File: test/JDBC/jdbc_schedule [hunk 9]
```diff
@@ -11,6 +11,13 @@
 all
 
 # Temporarily ignore tests until fixed in PG15
+ignore#!#BABEL-LOGIN-USER-EXT
```
Comment (surendravishnoi) [inline test/JDBC/jdbc_schedule:14]: why are these added as part of this PR?

File: test/JDBC/jdbc_schedule [hunk 10]
```diff
@@ -11,6 +11,13 @@
 all
 
 # Temporarily ignore tests until fixed in PG15
```
Comment (rishabhtanwar29) [inline test/JDBC/jdbc_schedule:13]: We should file a jira for these tests

**Review by rishabhtanwar29 — DISMISSED** [review]

File: test/JDBC/input/datatypes/TestVariableDataLength-vu-prepare.txt [hunk 11]
```diff
@@ -0,0 +1,42 @@
+# all variable length datatypes like numeric, decimal, time, datetime2, datetimeoffset need explicit checking
+# on their permissible datalength at the time of table or stored procedure creation and should generate error if required.
+
+# BABEL-3720
+
```
Comment (surendravishnoi) [inline test/JDBC/input/datatypes/TestVariableDataLength-vu-prepare.txt:5]: Have we validated all the results against SQL server?

File: test/JDBC/jdbc_schedule (hunk 10 above)
Comment (bsatarupa) [reply test/JDBC/jdbc_schedule:13]: Created: https://jira.rds.a2z.com/browse/BABEL-3811

File: test/JDBC/input/datatypes/TestVariableDataLength-vu-prepare.txt [hunk 12]
```diff
@@ -0,0 +1,42 @@
+# all variable length datatypes like numeric, decimal, time, datetime2, datetimeoffset need explicit checking
```
Comment (surendravishnoi) [inline test/JDBC/input/datatypes/TestVariableDataLength-vu-prepare.txt:1]: this change has both boundary and negative test cases. Let us update the description with relevant test details

File: test/JDBC/jdbc_schedule (hunk 9 above)
Comment (bsatarupa) [reply test/JDBC/jdbc_schedule:14]: I have created a new JIRA for fixing roles and permission related testcases on PG15 that are currently failing on JDBC framework https://jira.rds.a2z.com/browse/BABEL-3811

File: test/JDBC/input/datatypes/TestVariableDataLength-vu-prepare.txt (hunk 12 above)
Comment (bsatarupa) [reply test/JDBC/input/datatypes/TestVariableDataLength-vu-prepare.txt:1]: Done

File: test/JDBC/expected/BABEL-3486-vu-prepare.out (hunk 7 above)
Comment (bsatarupa) [reply test/JDBC/expected/BABEL-3486-vu-prepare.out:1062]: yes, SQL server error:
Started executing query at Line 45
Msg 1002, Level 15, State 1, Line 1
Line 1: Specified scale 10 is invalid.
Total execution time: 00:00:00.006

File: test/JDBC/input/datatypes/TestVariableDataLength-vu-prepare.txt (hunk 11 above)
Comment (bsatarupa) [reply test/JDBC/input/datatypes/TestVariableDataLength-vu-prepare.txt:5]: yes

File: test/JDBC/expected/BABEL-3614-vu-prepare.out (hunk 8 above)
Comment (bsatarupa) [reply test/JDBC/expected/BABEL-3614-vu-prepare.out:49]: Yes. SQL server error:
Started executing query at Line 47
Msg 1002, Level 15, State 1, Line 1
Line 1: Specified scale 80 is invalid.
Total execution time: 00:00:00.001

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 6 above)
Comment (bsatarupa) [reply contrib/babelfishpg_tsql/src/hooks.c]: done

Comment (bsatarupa) [reply contrib/babelfishpg_tsql/src/hooks.c]: used DeconstructQualifiedName() which returns typeName excluding schemaName as per our requirement.

**Review by surendravishnoi — APPROVED** [review]
